arangod/Aql/OptimizerRules.cpp

```diff
@@ -63,6 +63,37 @@ using namespace arangodb;
 using namespace arangodb::aql;
 using EN = arangodb::aql::ExecutionNode;
 
+namespace arangodb {
+namespace aql {
+
+// TODO cleanup this f-ing aql::Collection(s) mess
+Collection* addCollectionToQuery(Query* query, std::string const& cname, bool assert) {
+  aql::Collections* colls = query->collections();
+  aql::Collection* coll = colls->get(cname);
+
+  if (coll == nullptr && !cname.empty()) { // TODO: cleanup this mess
```

Comment:
Just to double-check, this is trying to identify the case where we have not found the collection by name in the query's registry/cache and the name is not empty, so we need to go look it up properly?

Reply:
This is some magic that is required because  `aql::Collection* coll = colls->get(cname);` does not find the collection even if you would expect it to. That function already existed and I just added a check for non-empty-name, because the `colls->add(...)` asserts a vaild name. For me a bad name could be ok, because it could be user supplied argument. In that case i just return a `nullptr` and fail a bit later.